WIP: Update Readme and kdocs (#22)
r0adkll/kimchi: README.md

# Kimchi

**K**_otlin_ **I**_nject_ **M**_erge_ **C**_omponent_ **H**_ints_ **I**_ntuitively_

> Kimchi is an [Anvil](https://github.com/square/anvil)-like KSP processor for [kotlin-inject](https://github.com/evant/kotlin-inject) that lets you contribute bindings, modules, and subcomponents across multi-module projects to form your dependency injection graph without having manual wiring your components upstream.

Reviewer: ```suggestion
Kimchi is an [Anvil](https://github.com/square/anvil)-like KSP processor for [kotlin-inject](https://github.com/evant/kotlin-inject) that lets you contribute bindings, modules, and subcomponents across multi-module projects to form your dependency injection graph without having manual to wire up your components upstream.
```

## Getting started

Getting started is easy! Just create a component abstract class or interface annotated with `@MergeComponent` like so

```kotlin
@MergeComponent(AppScope::class)
abstract class AppComponent
```

Now we have a component ready to collect contributions on the scope `AppScope`
Next, start contributing modules and bindings.

### Binding
```kotlin
interface MenuRepository {
  suspend fun getItems(): List<MenuItem>
}

@ContributesBinding(AppScope::class)
@Inject
class MenuRepositoryImpl(/*…*/) : MenuRepository {
  override suspend fun getItems(): List<MenuItem> = //…
}
```

### Multi-bindings
```kotlin
interface AppInitializer // { … }

@ContributesMultibinding(AppScope::class)
@Inject
class AnalyticsInitializer() : AppInitializer

@ContributesMultibinding(AppScope::class)
@Inject
class LoggingInitializer() : AppInitializer

// Will come together in a Set
@MergeComponent(AppScope::class)
abstract class AppComponent {
  abstract val initializers: Set<AppInitializer>
```

> or you can specify a `MapKey` (`StringKey`, `IntKey`, `LongKey`, or you can make a custom key if needed) for mapped bindings

Reviewer: this is clunky, but not sure how to improve. maybe something like

or you can specify a `MapKey` (`StringKey`, `IntKey`, `LongKey`, or custom key) for mapped bindings

  override fun create(userSession: UserSession): UserComponent =
    MergedUserComponent::class.create(userSession, this)

  @Component
  abstract class MergedUserComponent(
    @get:Provides val userSession: UserSession,
    @Component val parent: MergedAppComponent,
  ) : UserComponent // { … }
}

```

### Scopes

**Kimchi** uses scopes as markers to help the KSP processor connect the contributions you make to the target component you want to merge them to. The class `AppScope` from these examples can be represented like this:

```kotlin
object AppScope
```
These scope marker classes are independent of [kotlin-inject's scopes](https://github.com/evant/kotlin-inject?tab=readme-ov-file#scopes) but can be used together by creating a kotlin-inject scope wrapper like so:

```kotlin
// SingleIn.kt
@Scope // <-- kotlin-inject scoping
@Retention(AnnotationRetention.RUNTIME)
annotation class SingleIn(val scope: KClass<*>)

// AppComponent.kt
@SingleIn(AppScope::class)
@MergeComponent(AppScope::class)
abstract class AppComponent

// or…
@SingleIn(AppScope::class)
@ContributesBinding(AppScope::class)
class MenuRepositoryImpl : MenuRepository
```

### Exclusions & Replacements

> Components and bindings contributed with Kimchi can support replacements and overrides in a few ways. This is handy when you want to provide different bindings/contributions for instrumentation tests, or different build flavors.

Reviewer: ```suggestion
Components and bindings contributed with Kimchi can support replacements and overrides in a few ways. This is handy when you, for example, want to provide different bindings/contributions for instrumentation tests or different build flavors.
```

**Using `replaces = […]` parameter**

All `@Contributes*` annotations support specify an array of class references in their `replaces` parameter, like so:

```kotlin
@ContributesBinding(
  scope = AppScope::class,
  replaces = [MenuRepositoryImpl::class],
)
class TestMenuRepository : MenuRepository //…
```

> Then if both `TestMenuRepository` and `MenuRepositoryImpl` are on the classpath Kimchi will read the `replaces` argument and replace the bindings. This logic follows for all the other contribution annotations and merging.

Reviewer: ```suggestion
Then if both `TestMenuRepository` and `MenuRepositoryImpl` are on the classpath, Kimchi will read the `replaces` argument and replace the bindings. This logic follows for all the other contribution annotations and merging.
```

**Using `rank` parameter**

> When using `@ContributesBinding` if you don't have access to the class/implementation that you want to replace/override on the classpath (e.g. if you are doing a sort of `no-op` setup) then you can define a `rank` integer that kimchi will use to determine priority when merging  contributions of the same scope and bound on the classpath. The higher rank will always replace lower ranks and if components have the same rank then an error will be thrown.

Reviewer: ```suggestion
When using `@ContributesBinding`, if you don't have access to the class/implementation that you want to replace/override on the classpath (e.g. if you are doing a sort of `no-op` setup), then you can define a `rank` integer that kimchi will use to determine priority when merging  contributions of the same scope and bound on the classpath. The higher rank will always replace lower ranks. If components have the same rank then an error will be thrown.
```

e.g.
```kotlin
@ContributesBinding(
  scope = AppScope::class,
  rank = ContributesBinding.RANK_HIGHEST,
)
class NoOpMenuRepository : MenuRepository
```

> [!NOTE]   
> It is always recommended to use `replaces` instead of `rank` whenever possible. The `replaces` parameter will always take precedence no matter what you set the `rank` value to.

**Component Exclusions**

> You can explicitly declare contributed elements on the class path from being merged into components using the `excludes` parameter on the `@MergeComponent` and `@ContributesSubcomponent` annotations, like so:

Reviewer: ```suggestion
You can explicitly declare contributed elements on the classpath from being merged into components using the `excludes` parameter on the `@MergeComponent` and `@ContributesSubcomponent` annotations, like so:
```

```kotlin
@MergeComponent(
  scope = AppScope::class,
  excludes = [ SomeLegacyModule::class ],
)
abstract class AppComponent
```

> In a perfect world this feature shouldn't be needed. However, due to legacy setups, poor modularization, and other constraints of modern day software development applications might need to use it.

Reviewer: ```suggestion
In a perfect world this feature shouldn't be needed. However, due to legacy setups, poor modularization, and other constraints of modern day software development, applications might need to use it.
```
